Make typing for store generic

devceline opened the pull request:
# Description

- Make typing for store generic to make extending it easier

The store controller is generic and should not make assumptions on the object
types stored by it

## How Has This Been Tested?

2 new tests in `store.spec.ts` to check for retrieving from cached data and duplicate keys 

<!-- If valid for smoke test on feature add screenshots -->

## Due Dilligence

* [x] Breaking change
* [x] Requires a documentation update
* [x] Requires a e2e/integration test update

devceline committed:
Make typing for store generic

devceline committed:
Use record type for clarity

bkrem commented:
Commenting what I mentioned on Slack here for visibility, caveats:

- ~Store assumes that the value is effectively an object because it's using spread-merging here for update: https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/core/src/controllers/store.ts#L103~ Edit: Resolved by the fact that value stays as `Record<K,V>` here 👍 

- When restoring from persistence into memory, there's a check for whether `value.proposer.publicKey` or `value.topic` exists: https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/core/src/controllers/store.ts#L36

devceline committed:
Add getKey functionality and appropiate tests

This allows the store to presist data that is not a proposal or a
session, while being backwards compatible so that getKey is not required
for proposals or sessions

devceline committed:
Add null checking

xzilja commented on packages/core/src/controllers/store.ts:
```diff
@@ -40,6 +52,11 @@ export class Store<Key, Data extends StoreStruct> extends IStore<Key, Data> {
           // TODO(pedro) revert type casting as any
           this.map.set(value.topic as any, value);
+        } else if (this.getKey) {
+          // Null checking to prevent undefined or null values to throw an error
```
Could all of this be combined into `else if (this.getKey && value)` ? ideally also used alongside `isUndefined` util to prevent truthy values slipping by

bkrem commented on packages/core/src/controllers/store.ts:
```diff
@@ -1,30 +1,42 @@
+
   private storagePrefix = CORE_STORAGE_PREFIX;
 
+  /**
```
I think we can remove this, only makes sense to add these JSDoc style comments if we do it everywhere.

devceline replied:
Felt like it made sense specifically because `getKey` is not immediately straightforward.

Maybe I can remove the JSDoc and change the function's name to `getKeyForCache` or something, what do you think

bkrem replied:
Yeah assuming the type inference is meaningful enough when inspecting the constructor param, I think we can go with `getKeyForCache` or `getKeyForCacheRestore`.

If you feel like it's a lot better with the JSDoc comment feel free to leave it though 👍

devceline replied:
I personally feel like it makes sense to leave JSDoc in, maybe in the future `getKey` can be expanded on and used for other reasons (eg: Logging) so just adding some documentation doesn't hurt anyone I think. But I do not feel _very_ strongly about this.

bkrem replied:
Ok let's leave it there :)

xzilja reviewed (commented):
I assume this is needed to make store more reusable in other sdk's like chat and auth, correct?

bkrem commented on packages/core/src/controllers/store.ts:
```diff
@@ -1,30 +1,42 @@
     public name: string,
-    storagePrefix = CORE_STORAGE_PREFIX,
+    storagePrefix: string = CORE_STORAGE_PREFIX,
+    getKey: Store<Key, Data>["getKey"] = undefined,
```
Think this extra constructor param needs to be accounted for in `IStore`

devceline replied:
I didn't want to make the assumption that all `IStore` implementers will need to have this functionality. One could argue that one could have an `IStore` implementation that does not use the same caching mechanisms, no?

bkrem replied:
Yeah fair point, no need to add the assumption at this stage 👍

bkrem commented:
> I assume this is needed to make store more reusable in other sdk's like chat and auth, correct?

Exactly, at the moment the accepted values are opinionated towards Sign.

devceline committed:
Update null checking to be more precise

bkrem reviewed (approved):
Thank you for this @devceline 🙌 Nice work on the tests too 💯
